### PR TITLE
Make AppConfig client mock-able

### PR DESCRIPTION
**Pull request by pakrym**
Client SDK users need to be able to mock clients to be able to test their consumption code in isolation.

There are two main points of interaction here:
1. Client type itself
2. Models

To be able to mock the client type we have to adjust two things:
1. Make type constructable via protected ctor.
2. Make methods overrideable by marking them virtual

Model types are mostly POCO objects except for read-only properties that customers need a way to set to simulate real responses from servers.

There are multiple options on what we can do:
1. Provide ctor that would allow setting all the properties
2. Make properties virtual
3. Make setters protected

I went with the contructor option in PR for multiple reasons:

1. It has been used in existing SDKs and is familiar to customers
2. It's the easiest one to use - customers can just create instances of model classes instead. Protected setters require manually adding a type to be able to set properties. Virtual properties require using the mock object for every instance. I'll illustrate it bellow.

```
// Option 1

var mockSetting = new ConfigurationSetting("Key", "Value", "Label", locked: true);

// Option 2 (using Moq syntax)

var mock = new Mock<ConfigurationSetting>()
mock.SetupGet(m=>m.Locked).Returns(true);
var mockSetting = mock.Object;

// Option 3

var mockSetting = new MockConfigurationSetting("Key", "Value", "Label");
mockSetting.SetLocked(true);

class MockConfigurationSetting
{
    public MockConfigurationSetting(string key, string value, string label = null): base(key, value, label)
    {
    }

    public void SetLocked(bool value)
    {
         Locked = value;
    }
}

```

When this PR goes in I'll add requirements we decide on to the dotnet guidelines.

### CODE REVIEW

File: src/SDKs/Azure.ApplicationModel.Configuration/data-plane/Azure.ApplicationModel.Configuration/ConfigurationSetting.cs [hunk 1]
```diff
@@ -72,20 +73,8 @@ public ConfigurationSetting(string key, string value, string label = null)
         /// A dictionary of tags that can help identify what a key-value may be applicable for.
         /// </summary>
         public IDictionary<string, string> Tags {
-            get {
-                if (_tags == null) {
-                    lock (_key) {
```
Comment (pakrym) [inline src/SDKs/Azure.ApplicationModel.Configuration/data-plane/Azure.ApplicationModel.Configuration/ConfigurationSetting.cs]: Model types are not required to be thread-safe but locking on string instances is not the best thing to do in any case. In addition, while Tags getter might be made thread safe using the lock `Dictionary<T,T>` is not threadsafe and concurrent access may cause its data structure corruption and lead to infinite loops.

Comment (jsquire) [reply src/SDKs/Azure.ApplicationModel.Configuration/data-plane/Azure.ApplicationModel.Configuration/ConfigurationSetting.cs]: For what it's worth,  +1 for removing the lock.   If we really need to make this thread safe for instantiation, my vote goes to using `Lazy<T>` over a lock.

**Review by jsquire — APPROVED** [review]
I'm a strong vote for approach #1, personally.    

Though, I would probably favor using an interface abstraction for the client, such as an `IConfigurationClient` over virtualizing the mthods.  The interface would still allow for a mocking framework or a test-specific implementation to control behavior for testing, but without potentially triggering network calls from the the base implementation by accident..

**Comment by pakrym** [comment]
> Though, I would probably favor using an interface abstraction for the client, such as an IConfigurationClient over virtualizing the mthods. 

You can't add members to interfaces without it being a breaking change so I'm less sure about this.

**Comment by jsquire** [comment]
> > Though, I would probably favor using an interface abstraction for the client, such as an IConfigurationClient over virtualizing the mthods.
> 
> You can't add members to interfaces without it being a breaking change so I'm less sure about this.

True, but if we're changing the contract, I'd want my tests to know.   Assuming a decent mocking framework is in use, it would automatically create the stubs based on the new contract version.   The only scenario that may be potentially an issue is creating a test-specific implementation of the interface which would, as noted, break.

**Comment by pakrym** [comment]
> True, but if we're changing the contract, I'd want my tests to know. Assuming a decent mocking framework is in use, it would automatically create the stubs based on the new contract version. The only scenario that may be potentially an issue is creating a test-specific implementation of the interface which would, as noted, break.

But most of the time you wouldn't care, adding a new member to the client rarely changes what your tests did. 

> Assuming a decent mocking framework is in use, it would automatically create the stubs based on the new contract version. The only scenario that may be potentially an issue is creating a test-specific implementation of the interface which would, as noted, break.

True, but if the interface ships in the product assembly we can't know if it's only used by tests.

**Comment by jsquire** [comment]
> But most of the time you wouldn't care, adding a new member to the client rarely changes what your tests did.

Which is why mocking frameworks are used widely.

> True, but if the interface ships in the product assembly we can't know if it's only used by tests.

Sure, but do you see that as a common scenario?  Creating a specialized version of the client outside of testing?   

Pure addition is fine, but if we're potentially changing behavior as part of the iteration, then we've still got the question of how to version.   In my experience, outside of additions, maintaining backwards compatibility through non-trivial changes to a service is hard.  It seldom goes as easily as we all want it to.   API surface changes to a service typically require a new contract.  Either we're adding an additional interface or we're adding additional methods that potentially pollute the surface area of the class, making it more difficult to discover things.   

To be clear, I've no strong objection to the approach that you've taken.  My personal preference is to drive off of interfaces where possible and, in the case where we don't want to introduce a breaking change then we introduce a new interface, even if the implementation lives on the existing class, you get a clear segmentation of the API surface.

File: src/SDKs/Azure.ApplicationModel.Configuration/data-plane/Azure.ApplicationModel.Configuration/ConfigurationClient.cs [hunk 2]
```diff
@@ -17,6 +17,10 @@ public partial class ConfigurationClient
         private readonly Uri _baseUri;
         private readonly HttpPipeline _pipeline;
 
+        protected ConfigurationClient()
```
Comment (KrzysztofCwalina) [inline src/SDKs/Azure.ApplicationModel.Configuration/data-plane/Azure.ApplicationModel.Configuration/ConfigurationClient.cs]: Can you add doc comments saying that this is for mocking scenarios

File: src/SDKs/Azure.ApplicationModel.Configuration/data-plane/Azure.ApplicationModel.Configuration/ConfigurationSetting.cs [hunk 3]
```diff
@@ -23,17 +22,19 @@ public ConfigurationSetting(string key, string value, string label = null)
             Label = label;
         }
 
+        public ConfigurationSetting(string key, string value, string label, string contentType = null, ETag eTag = default, DateTimeOffset? lastModified = null, bool? locked = null) : this(key, value, label)
```
Comment (KrzysztofCwalina) [inline src/SDKs/Azure.ApplicationModel.Configuration/data-plane/Azure.ApplicationModel.Configuration/ConfigurationSetting.cs]: Yeah, I really don't like this. It complicates the mainline API. In particular I don't like that now ETag is settable. I would prefer that such APIs shipped in either a test specific assembly, or at least test namespace/type. In the worst case, we should consider making the APIs not visible in intellisense by either making the setters protected, or applying the visibility attributes

File: src/SDKs/Azure.ApplicationModel.Configuration/data-plane/Azure.ApplicationModel.Configuration/ConfigurationSetting.cs [hunk 4]
```diff
@@ -23,17 +22,19 @@ public ConfigurationSetting(string key, string value, string label = null)
...
         /// The key is used in unison with the label to uniquely identify a key-value.
         /// </summary>
-        public string Key {
-            get => _key;
-            set {
-                if (value == null) throw new ArgumentNullException(nameof(Key));
-                _key = value;
-            }
-        }
+        public string Key { get; set; }
```
Comment (KrzysztofCwalina) [inline src/SDKs/Azure.ApplicationModel.Configuration/data-plane/Azure.ApplicationModel.Configuration/ConfigurationSetting.cs]: Why this change?

**Review by KrzysztofCwalina — COMMENTED** [review]
Let's chat how we can make the model mutating APIs not so visible.

**Comment by dsgouda** [comment]
@pakrym Please pull from latest master and update the PR to fix the Travis failures

File: src/SDKs/Azure.ApplicationModel.Configuration/data-plane/Azure.ApplicationModel.Configuration/ConfigurationClient.cs (hunk 2 above)
Comment (pakrym) [reply src/SDKs/Azure.ApplicationModel.Configuration/data-plane/Azure.ApplicationModel.Configuration/ConfigurationClient.cs]: Will do.

File: src/SDKs/Azure.ApplicationModel.Configuration/data-plane/Azure.ApplicationModel.Configuration/ConfigurationSetting.cs (hunk 4 above)
Comment (pakrym) [reply src/SDKs/Azure.ApplicationModel.Configuration/data-plane/Azure.ApplicationModel.Configuration/ConfigurationSetting.cs]: Having validator in this property setter is pretty random. For example, Value is required too and is not being validated. I think we should validate in service calls instead of model setter because required values might be dependent on the context object is used in.

The other thing to think about here is nullable annotations, it might slightly change how we design model types.

File: src/SDKs/Azure.ApplicationModel.Configuration/data-plane/Azure.ApplicationModel.Configuration/ConfigurationClientModelFactory.cs [hunk 5]
```diff
@@ -0,0 +1,33 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+
+namespace Azure.ApplicationModel.Configuration
+{
+    public static class ConfigurationClientModelFactory
```
Comment (KrzysztofCwalina) [inline src/SDKs/Azure.ApplicationModel.Configuration/data-plane/Azure.ApplicationModel.Configuration/ConfigurationClientModelFactory.cs]: Should we simply call these types ModelBuilder? i.e. each namespace would have a type with this name?

File: src/SDKs/Azure.ApplicationModel.Configuration/data-plane/Azure.ApplicationModel.Configuration/ConfigurationClientModelFactory.cs [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+    {
+        public static ConfigurationSetting ConfigurationSetting(
```
Comment (KrzysztofCwalina) [inline src/SDKs/Azure.ApplicationModel.Configuration/data-plane/Azure.ApplicationModel.Configuration/ConfigurationClientModelFactory.cs]: Should these be called Create<return_type>?

**Review by KrzysztofCwalina — APPROVED** [review]

File: src/SDKs/Azure.ApplicationModel.Configuration/data-plane/Azure.ApplicationModel.Configuration/ConfigurationClientModelFactory.cs (hunk 6 above)
Comment (pakrym) [reply src/SDKs/Azure.ApplicationModel.Configuration/data-plane/Azure.ApplicationModel.Configuration/ConfigurationClientModelFactory.cs]: I followed Roslyns `SyntaxFactory` (https://docs.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.csharp.syntaxfactory?view=roslyn-dotnet) pattern but have no strong opinion.

File: src/SDKs/Azure.ApplicationModel.Configuration/data-plane/Azure.ApplicationModel.Configuration/ConfigurationClientModelFactory.cs (hunk 5 above)
Comment (pakrym) [reply src/SDKs/Azure.ApplicationModel.Configuration/data-plane/Azure.ApplicationModel.Configuration/ConfigurationClientModelFactory.cs]: This might be fine for App Config but I can see multiple storage clients being used in the same source file. I think we should avoid having types with the same name in multiple SKDs as a rule.

File: src/SDKs/Azure.ApplicationModel.Configuration/data-plane/Azure.ApplicationModel.Configuration/ConfigurationClientModelFactory.cs (hunk 6 above)
Comment (pakrym) [reply src/SDKs/Azure.ApplicationModel.Configuration/data-plane/Azure.ApplicationModel.Configuration/ConfigurationClientModelFactory.cs]: The type name was slightly long so I went with shorter method names.
